numenc/encoder_decoder.c

```diff
@@ -0,0 +1,947 @@
+    }
+
+    union double_bytes f_bytes;
+    f_bytes.val = input;
+
+    char * buffer = (char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    // positive number: set sign bit to 1
+    if (input >= 0) {
+        if (endianness == 0) {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[i] = f_bytes.bytes[i];
+            }
+        } else {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[EIGHT_BYTES - 1 - i] = f_bytes.bytes[i];
+            }
+        }
+        buffer[0] |= 0x80;
+        // negative number: flip all bits
+    } else {
+        if (endianness == 0) {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[i] = ~f_bytes.bytes[i];
+            }
+        } else {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[EIGHT_BYTES - 1 - i] = ~f_bytes.bytes[i];
+            }
+        }
+    }
+
+    output = PyBytes_FromStringAndSize(buffer, EIGHT_BYTES);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyObject * to_float64(PyObject * self, PyObject * args) {
+    const char * input;
+    PyObject * output;
+    int count;
+
+    if (!PyArg_ParseTuple(args, "y#", & input, & count)) {
+        return PyErr_Format(PyExc_TypeError, "Wrong input: expected bytes.");
+    }
+
+    if (endianness == -1) {
+        endianness = endianness_impl();
+    }
+
+    if (count != EIGHT_BYTES) {
+        return PyErr_Format(PyExc_ValueError, "Illegal input: expected bytes of length %d, got %d.",
+                            EIGHT_BYTES, count);
+    }
+
+    unsigned char * buffer = (unsigned char * ) PyMem_RawMalloc(EIGHT_BYTES);
+    if (buffer == NULL) {
+        return PyErr_NoMemory();
+    }
+
+    // sign bit is 1: positive number
+    if (input[0] & 0x80) {
+        if (endianness == 0) {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[i] = (unsigned char) input[i];
+            }
+            buffer[0] ^= 0x80;
+        } else {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[EIGHT_BYTES - 1 - i] = (unsigned char) input[i];
+            }
+            buffer[EIGHT_BYTES - 1] ^= 0x80;
+        }
+    } else {
+        if (endianness == 0) {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[i] = ~((unsigned char) input[i]);
+            }
+        } else {
+            for (int i = 0; i < EIGHT_BYTES; i++) {
+                buffer[EIGHT_BYTES - 1 - i] = ~((unsigned char) input[i]);
+            }
+        }
+    }
+
+    union double_bytes d_bytes;
+
+    for (int i = 0; i < EIGHT_BYTES; i++) {
+        d_bytes.bytes[i] = buffer[i];
+    }
+
+    output = Py_BuildValue("d", d_bytes.val);
+    PyMem_RawFree(buffer);
+    return output;
+}
+
+static PyMethodDef EncdecMethods[] = {
+    {
+        "from_int8",
+        from_int8,
+        METH_VARARGS,
```

Comment:
Please specify also the function signature in the docstring (see what is the achievable maximum in: https://stackoverflow.com/questions/50537407/add-a-signature-with-annotations-to-extension-methods).

Reply:
According to that link, it's "maybe possible, but probably very complicated and almost certainly not worth the trouble" to add signatures to C extension modules. The most viable solution it purposes is to wrap C code in the thinnest python script wrapper possible - but that'd defy the "speed" purpose of the library.